editing/edit-with-branch-versioning/src/main/java/com/esri/samples/edit_with_branch_versioning/EditWithBranchVersioningController.java

```diff
@@ -0,0 +1,378 @@
+        switchVersion(userCreatedVersionName);
+
+        // hide the form from the UI as the sample only allows 1 version to be created
+        createVersionVBox.setVisible(false);
+        switchVersionButton.setDisable(false);
+      } catch (Exception ex) {
+        // if there is an error creating a new version, display an alert and reset the UI
+        if (ex.getCause().toString().contains("The version already exists")) {
+          showAlert("A version with this name already exists.\\nPlease enter a unique name");
+        } else {
+          showAlert("Error creating new version");
+        }
+        createVersionButton.setText("Create version");
+        createVersionButton.setDisable(false);
+      }
+    });
+  }
+
+  /**
+   * Apply local edits to the service geodatabase and switch branch version.
+   */
+  @FXML
+  private void handleSwitchVersionButtonClicked() {
+
+    if (serviceGeodatabase.getVersionName().equals(defaultVersionName)) {
+      // if the current version is the default version, switch to the user created version
+      switchVersion(userCreatedVersionName);
+
+    } else if (serviceGeodatabase.getVersionName().equals(userCreatedVersionName)) {
+      // if the current version is the user created version, check if there are local edits
+      if (!serviceGeodatabase.hasLocalEdits()) {
+        // if there are no local edits switch to the default version
+        switchVersion(defaultVersionName);
+
+      } else {
+        // if local edits exist apply the edits
+        ListenableFuture<List<FeatureTableEditResult>> resultOfApplyEdits = serviceGeodatabase.applyEditsAsync();
+        resultOfApplyEdits.addDoneListener(() -> {
+          try {
+
+            // check if the server edit was successful
+            List<FeatureTableEditResult> edits = resultOfApplyEdits.get();
+            if (edits == null || edits.isEmpty()) {
+              showAlert("Error applying edits on server");
+            } else {
+              // if the edits were successful, switch to the default version
+              switchVersion(defaultVersionName);
+            }
+          } catch (InterruptedException | ExecutionException e) {
+            showAlert("Error applying edits on server");
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * Switch the active branch version.
+   *
+   * @param versionName name of the version to switch to
+   */
+  private void switchVersion(String versionName) {
+    ListenableFuture<Void> switchVersionFuture = serviceGeodatabase.switchVersionAsync(versionName);
+    switchVersionFuture.addDoneListener(() -> {
+      // check if the active version has switched successfully and update the UI
+      if (serviceGeodatabase.getVersionName().equals(versionName)) {
+        currentVersionLabel.setText("Current version: " + serviceGeodatabase.getVersionName());
+        editFeatureVBox.setDisable(true);
+      } else {
+        showAlert("Error switching version");
+      }
+    });
```

Comment:
having a look at some similar samples which use `ListenableFuture`s, I'm recognising the use of a try/catch within their addDoneListeners. There are some instances in the code where this `switchVersion` method is called within a try/catch anyway, so it's caught there, but in some places it isn't. We will get a more meaningful error message for the user by catching an exception and printing it out rather than a hard coded string alert: although this can be helpful too. 

```suggestion
    switchVersionFuture.addDoneListener(() -> {
      try {
        // check if the active version has switched successfully and update the UI
        if (serviceGeodatabase.getVersionName().equals(versionName)) {
          currentVersionLabel.setText("Current version: " + serviceGeodatabase.getVersionName());
          editFeatureVBox.setDisable(true);
        }
      } catch (Exception e) {
        e.printStackTrace();
      }
    });
```

Reply:
Added this in. The other place with a listenable futures was the update features method (line 318) but in other samples this doesn't seem to have a try catch, so I've left it as it is with the if statement and alert. I guess any exceptions would get caught in the initialize try/catch when that method is called. Let me know if you think this is justified or should be added.